context/app/api/client.py

```diff
@@ -165,45 +238,99 @@ def get_provenance(self, uuid):
-                },
-            ]
+                ],
+                'name': 'Linnarsson',
+                'staticLayout': [
+                    {
+                        'component': 'scatterplot',
+                        'props': {
+                            'mapping': 'UMAP',
+                            'view': {
+                                'zoom': 4,
+                                'target': [0, 0, 0]
+                            }
+                        },
+                        'x': 0, 'y': 0, 'w': 12, 'h': 2
+                    },
+                ]
+            }
+        else:
+            return self._build_vitessce_conf(entity)
+
+    def _build_assets_url(self, rel_path, uuid):
+        base_url = urllib.parse.urljoin(
+            current_app.config['ASSETS_ENDPOINT'],
+            str(Path(uuid) / Path(rel_path))
+        )
+        token_param = urllib.parse.urlencode({'token': self.nexus_token})
+        return base_url + '?' + token_param
+
+    def _build_image_schema(self, rel_path, uuid):
+        schema = {}
+        schema['name'] = Path(rel_path).name
+        schema['type'] = 'ome-tiff'
+        schema['url'] = self._build_assets_url(rel_path, uuid)
+        schema['metadata'] = {}
+        offsets_path = Path(CODEX_OFFSETS_PATH) / \
+            Path(Path(rel_path).name.replace('ome.tiff', 'offsets.json'))
+        schema['metadata']['omeTiffOffsetsUrl'] = self._build_assets_url(str(offsets_path), uuid)
+        return schema
+
+    def _build_image_layer_datauri(self, rel_path, uuid):
+        image_layer = {}
+        image_paths = [
+            str(rel_path).replace(
+                '#CODEX_TILE#', f'R{str(r).zfill(3)}_X{str(x).zfill(3)}_Y{str(y).zfill(3)}'
+            ) for (r, x, y) in itertools.product(*[R_VALS, X_VALS, Y_VALS])
+        ]
+        image_layer['images'] = [
+            self._build_image_schema(image_path, uuid) for image_path in image_paths
+        ]
+        image_layer['schema_version'] = '0.0.1'
+        return DataURI.make(
+            'text/plain', charset='us-ascii', base64=True, data=json.dumps(image_layer)
+        )
+
+    def _build_layer_conf(self, file, uuid, assay_type):
+        return {
+            'type': file['type'],
+            'url': self._build_assets_url(file['rel_path'], uuid)
+            if assay_type not in IMAGE_ASSAYS
+            else self._build_image_layer_datauri(file['rel_path'], uuid),
+            'name': file['type'].lower(),
         }
+
+    def _build_vitessce_conf(self, entity):
+        # Can there be more than one of these?  This seems like a fine default for now.
+        assay_type = entity['data_types'][0]
+        uuid = entity['uuid']
+        conf = ASSAY_CONF_LOOKUP[assay_type]['base_conf']
+        files = ASSAY_CONF_LOOKUP[assay_type]['files_conf']
+        layers = [self._build_layer_conf(file, uuid, assay_type) for file in files]
+
+        conf['layers'] = layers
+        conf['name'] = uuid
+        return conf
```

Comment:
Can you pull all this new code into a file by itself, and maybe there's only one public function?

And can you add doctests on these functions? Shouldn't be hard to write, and it would help to understand what each part does.